base_layer/key_manager/src/birthday.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright 2021. The Tari Project
+//
+// Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use std::convert::TryFrom;
+
+use serde::{Deserialize, Serialize};
+use tari_common::configuration::Network;
+use tari_core::blocks::genesis_block::get_genesis_block;
+
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Birthday {
+    zero_point_time: u64,
+    birthday: u16,
+    version: u8,
+}
+
+impl Birthday {
+    pub fn default() -> Self {
+        let network = Network::Dibbler;
+        Self::new(network)
+    }
+
+    pub fn new(network: Network) -> Self {
+        let current_time = Self::current_time_in_seconds();
+        Self::new_from_current_time(network, current_time)
+    }
+
+    fn new_from_current_time(network: Network, current_time: u64) -> Self {
+        const SECONDS_PER_DAY: u64 = 24 * 60 * 60;
+        const PERIOD_LENGTH: u64 = u16::MAX as u64 + 1; // 2^16
+
+        let mut zero_point_time = Self::get_network_genesis_time(network);
+
+        let days = (current_time - zero_point_time) / SECONDS_PER_DAY;
+        let birthday = u16::try_from(days % PERIOD_LENGTH).unwrap();
+        let version = u8::try_from(days / PERIOD_LENGTH).unwrap();
```

Comment:
When I implemented the code, I had in mind the third suggestion described [here](https://github.com/tari-project/tari/issues/4176). 

From what I understood, the `version` field in `Birthday` should really work as an epoch counter. This will be used later on to index the `version` in `CipherSeed`. Following this approach, versioning is epoch-dependent.

I am going to add docs, but do you think this approach makes sense  ?

Reply:
If one decides to keep the versioning for `CipherSeed` to be dependent on the epoch. Another approach would be to migrate the version computation to the `CipherSeed` impl, thus make it more explicit the use case we have in mind.

Reply:
The intent of the overall `version` field should be an unambiguous identification of all parts of the protocol that all software agrees on. So if you were to change `Argon2` parameters or migrate to a different MAC algorithm, you'd define a new version; any software that knows about it will do the right thing, and any software that doesn't will simply produce an error.

I don't, however, think it's a good idea to attempt to compute a dynamic version number from the birthday. The intent of the comment in [https://github.com/tari-project/tari/issues/4176](this issue) was that once enough time passes such that a given birthday could potentially overflow, then you (or whoever is doing development that far into the future!) would manually define a new version that accounts for this. Doing it dynamically feels like a risky approach that could cause conflicts with any non-birthday protocol changes requiring new versions.